src/core/buffer/events_generators.ts

```diff
@@ -0,0 +1,198 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Subject } from "rxjs";
+import { ICustomError } from "../../errors";
+import {
+  Adaptation,
+  ISegment,
+  Period,
+  Representation,
+} from "../../manifest";
+import { IBufferType } from "../source_buffers";
+import {
+  IActivePeriodChangedEvent,
+  IAdaptationChangeEvent,
+  IBitrateEstimationChangeEvent,
+  IBufferEventAddedSegment,
+  IBufferNeedsDiscontinuitySeek,
+  IBufferNeedsManifestRefresh,
+  IBufferStateActive,
+  IBufferStateFull,
+  IBufferWarningEvent,
+  ICompletedBufferEvent,
+  IEndOfStreamEvent,
+  IPeriodBufferClearedEvent,
+  IPeriodBufferReadyEvent,
+  IRepresentationChangeEvent,
+  IResumeStreamEvent,
+} from "./types";
+
+const EVENTS = {
+  activeBuffer(bufferType: IBufferType) : IBufferStateActive {
+    return {
+      type: "active-buffer",
+      value: { bufferType },
+    };
+  },
+
+  activePeriodChanged(period : Period) : IActivePeriodChangedEvent {
+    return {
+      type : "activePeriodChanged",
+      value : {
+        period,
+      },
+    };
+  },
+
+  adaptationChange(
+    bufferType : IBufferType,
+    adaptation : Adaptation|null,
+    period : Period
+  ) : IAdaptationChangeEvent {
+    return {
+      type: "adaptationChange",
+      value : {
+        type: bufferType,
+        adaptation,
+        period,
+      },
+    };
+  },
+
+  addedSegment<T>(
+    bufferType : IBufferType,
+    segment : ISegment,
+    segmentData : T
+  ) : IBufferEventAddedSegment<T> {
+    return {
+      type : "added-segment",
+      value : { bufferType, segment, segmentData },
+    };
+  },
+
+  bitrateEstimationChange(
+    type : IBufferType,
+    bitrate : number|undefined
+  ) : IBitrateEstimationChangeEvent {
+    return {
+      type: "bitrateEstimationChange",
+      value: {
+        type,
+        bitrate,
+      },
+    };
+  },
+
+  bufferComplete(bufferType: IBufferType) : ICompletedBufferEvent {
+    return {
+      type: "complete-buffer",
+      value: {
+        type: bufferType,
+      },
+    };
+  },
+
+  discontinuityEncountered(
+    bufferType : IBufferType,
+    nextTime : number
+  ) : IBufferNeedsDiscontinuitySeek {
+    return {
+      type : "discontinuity-encountered",
+      value : { bufferType, nextTime },
+    };
+  },
+
+  endOfStream() : IEndOfStreamEvent {
+    return {
+      type: "end-of-stream",
+      value: undefined,
+    };
+  },
+
+  fullBuffer(bufferType : IBufferType) : IBufferStateFull {
+    return {
+      type: "full-buffer",
+      value: { bufferType },
+    };
+  },
+
+  needsManifestRefresh(bufferType : IBufferType) : IBufferNeedsManifestRefresh {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Update or refactor this function so that its implementation doesn't duplicate the one on line 45. ([typescript:S4144](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4144))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=283&issues=AWUoCSgDOTG07Bqb98zF&open=AWUoCSgDOTG07Bqb98zF)

[ISSUE_KEY]: AWUoCSgDOTG07Bqb98zF